core/services/headtracker/head_listener.go

```diff
@@ -29,64 +28,48 @@ var (
+//go:generate mockery --name HeadListener --output ./mocks/ --case=underscore
 
-// Config configures headtracker and related structs
-type Config interface {
-	BlockEmissionIdleWarningThreshold() time.Duration
-	EvmFinalityDepth() uint32
-	EvmHeadTrackerHistoryDepth() uint32
-	EvmHeadTrackerMaxBufferSize() uint32
-	EvmHeadTrackerSamplingInterval() time.Duration
+// NewHeadHandler is a callback that handles incoming heads
+type NewHeadHandler func(ctx context.Context, header *eth.Head) error
+
+// HeadListener manages eth.Client connection that receives heads from the eth node
+type HeadListener interface {
+	// ListenForNewHeads kicks off the listen loop (not thread safe)
+	// done() must be executed upon leaving ListenForNewHeads()
+	ListenForNewHeads(handleNewHead NewHeadHandler, done func())
+	// ReceivingHeads returns true if the listener is receiving heads (thread safe)
+	ReceivingHeads() bool
+	// Connected returns true if the listener is connected (thread safe)
+	Connected() bool
 }
 
-// HeadListener manages the websocket connection that receives heads from the
-// eth node
-type HeadListener struct {
-	config                Config
-	ethClient             eth.Client
-	chainID               big.Int
-	headers               chan *eth.Head
-	headSubscription      ethereum.Subscription
-	connectedMutex        sync.RWMutex
-	connected             bool
-	receivesHeads         atomic.Bool
-	subscribeRetryBackoff backoff.Backoff
-
-	log logger.Logger
-
-	chStop chan struct{}
+type headListener struct {
+	config           Config
+	ethClient        eth.Client
+	logger           logger.Logger
+	stopCh           chan struct{}
+	headersCh        chan *eth.Head
+	headSubscription ethereum.Subscription
+	connected        atomic.Bool
+	receivingHeads   atomic.Bool
 }
 
 // NewHeadListener creates a new HeadListener
-func NewHeadListener(l logger.Logger,
-	ethClient eth.Client,
-	config Config,
-	chStop chan struct{},
-) *HeadListener {
-	if ethClient == nil {
-		panic("head listener requires non-nil ethclient")
-	}
-	return &HeadListener{
+func NewHeadListener(logger logger.Logger, ethClient eth.Client, config Config, stopCh chan struct{}) HeadListener {
+	return &headListener{
 		config:    config,
 		ethClient: ethClient,
-		chainID:   *ethClient.ChainID(),
-		subscribeRetryBackoff: backoff.Backoff{
-			Min: 1 * time.Second,
-			Max: 10 * time.Second,
-		},
-		log:    l.Named("listener"),
-		chStop: chStop,
+		logger:    logger.Named("listener"),
+		stopCh:    stopCh,
 	}
 }
 
-// ListenForNewHeads kicks off the listen loop
-// NOT THREAD SAFE
-func (hl *HeadListener) ListenForNewHeads(handleNewHead func(ctx context.Context, header *eth.Head) error, done func()) {
+func (hl *headListener) ListenForNewHeads(handleNewHead NewHeadHandler, done func()) {
 	defer done()
-	defer hl.unsubscribeFromHead()
+	defer hl.unsubscribe()
 
-	ctx, cancel := utils.ContextFromChan(hl.chStop)
+	ctx, cancel := utils.ContextFromChan(hl.stopCh)
```

Comment:
Extremely minor point but generally we prefix channels with `ch` rather than postfix.